### DISCUSSION TITLE
create addon without required options

### DISCUSSION THREAD

**Discussion by CarlosGrohmann** [OSGeo/grass#1944] [Q&A] [answered]
Hi

I'm working on an addon for burn area indices (NBR, dNBR, RdNBR). The thing is, all the options might be optional, and if I set all to `#% required : no`, when I call the module in the CLI, it tries to run without opening the GUI (since nothing is required).

So, is there a way to make it open the GUI when all the options are required?

thanks

**Comment by neteler** [comment, marked as answer]
You mean like with `--ui`?

**Reply by CarlosGrohmann** [reply]
Yes, kinda. 
Because if one installs it with g.extension, calling it in CLI would require `--ui` from the user as well, right? Which is not what an end-user would expect. Speaking of which, how to add addons to the GUI menu?

**Reply by marisn** [reply]
> Yes, kinda. Because if one installs it with g.extension, calling it in CLI would require `--ui` from the user as well, right? Which is not what an end-user would expect. 

This affects some core modules too (g.rename). It was a deliberate decision. I can't find a discussion about it now, but I remember my complaining about this "user friendly" new behaviour.

**Reply by neteler** [reply]
> > Yes, kinda. Because if one installs it with g.extension, calling it in CLI would require `--ui` from the user as well, right? Which is not what an end-user would expect.
> 
> This affects some core modules too (g.rename). It was a deliberate decision. I can't find a discussion about it now, but I remember my complaining about this "user friendly" new behaviour.

A message candidate :-) might be this one:

"[GRASS-dev] why doesn't g.region launch gui?"
https://lists.osgeo.org/pipermail/grass-dev/2012-February/058013.html

**Reply by marisn** [reply]
> > > Yes, kinda. Because if one installs it with g.extension, calling it in CLI would require `--ui` from the user as well, right? Which is not what an end-user would expect.
> > 
> > 
> > This affects some core modules too (g.rename). It was a deliberate decision. I can't find a discussion about it now, but I remember my complaining about this "user friendly" new behaviour.
> 
> A message candidate :-) might be this one:
> 
> "[GRASS-dev] why doesn't g.region launch gui?" https://lists.osgeo.org/pipermail/grass-dev/2012-February/058013.html

Thanks, Markus. Here is the real message:
https://grass-dev.osgeo.narkive.com/L36m4FkN/wxgui-new-packages-layout#post15

> Glynn Clements:
>  Originally, G_parser() generated the UI if a module was run with no
> arguments. Individual modules had to explicitly skip the G_parser()
> call if they could meaningfully be run without arguments.
> 
> However, skipping G_parser() altogether has side-effects which were
> frequently overlooked, e.g. setting the program name from argv[0],
> parsing opt->options into opt->opts[], parsing opt->descriptions into
> opt->descs[], etc.
> 
> Consequently, a decision was made to always call G_parser(), and to
> only generate a UI when no arguments are given and at least one option
> is required (to do otherwise would be a nuisance for modules which can
> be and often are run without options).
> 
> This falls down in the case where at least one option is required, but
> not any particular option. Modules can (and should) generate an error
> in this case, but currently there's no way to instruct the parser to
> produce a GUI.
> 
> Well, there's no "proper" way to do this, although it would be
> possible to "hack" the argc and argv which are passed to G_parser(),
> e.g.:
> 
> if (argc < 2) {
> static char *args[3];
> args[0] = argv[0];
> args[1] = "--ui";
> args[2] = NULL;
> argv = args;
> argc = 2;
> }
> if (G_parser(argc, argv))
> exit(EXIT_FAILURE);
> 
> But really we need to come up with a long-term solution to the
> situation where a module has requirements beyond the simplest case of
> individual options being required. A solution should ideally handle
> the "at least one of" case, as well as mutual exclusion and
> dependencies (if A is given, B must also be given). It should also be
> able to generate meaningful error messages (i.e. not just "that
> combination of options is invalid").

**Reply by CarlosGrohmann** [reply]
Very interesting discussion!

Is there a way to force the module to open the GUI internally? For example, `if len(options)==0`....?

**Reply by marisn** [reply]
As Glynn wrote – just fiddle with args before calling G_parser().

**Reply by CarlosGrohmann** [reply]
But that is in C, what about a Python way?

**Reply by marisn** [reply]
The same. Hack around with sys.argv.append() before calling grass.parser(). This is a nasty way and a proper solution in parser is needed.

**Reply by CarlosGrohmann** [reply]
thanks.

this works
```
if __name__ == "__main__":
    argv = sys.argv[:]
    if len(argv)<2:
        sys.argv.append('--ui')
    options, flags = gscript.parser()
    main()
```

**Comment by wenzeslaus** [comment]
As far as I can tell, the discussion referenced in the `--ui` thread happened before introduction of an additional [rules](https://grass.osgeo.org/grass78/manuals/g.parser.html#conditional-parameters) section (`rules:...required:...`) into the module/tool interface description used by the parser. Now the parser can tell if there are optional parameters, but at least one of them is required. It is not clear to me if that's applicable to @CarlosGrohmann's code, but notably `g.region` starts GUI and it uses _G_option_required_ (in C) while running `g.rename` is NOP and it does _not_ have any additional rules specified.
